### PR TITLE
PEP-563

### PR DESCRIPTION
**Pull request by da-woods**
Annotations are now dealt with according to PEP-563 - there are
saved as strings, rather than evaluated as Python objects.
They can/are still be used by Cython for typing.
Previous behaviour for evaluating them as Python objects was
convoluted and has been removed entirely, which hopefully doesn't
break too much.

------------------

I'd created a previous version of this that attempted to preserve the old behaviour too. Upon thinking about it a bit more, most of the messy code was just trying to make the pre-PEP-563 behaviour work, so I ditched it (like I should have done from the start).

I believe this should fix the test `embedsignatures` from the enablebinding branch too. It might be worth copying a few of my extra tests from the PR there, but that can be looked at separately (if it's worthwhile at all)

### CODE REVIEW

File: Cython/Compiler/Parsing.py [hunk 1]
```diff
@@ -3865,3 +3865,12 @@ def print_parse_tree(f, node, level, key = None):
             f.write("%s]\n" % ind)
             return
     f.write("%s%s\n" % (ind, node))
+
+def p_annotation(s):
+    """An annotation just has the "test" syntax, but also stores the string it came from
+
+    Note that the string is *allowed* to be changed/processed (although isn't here)
+    so may not exactly match the string generated by Python, and if it doesn't
+    then it is not a bug"""
```
Comment (scoder) [inline Cython/Compiler/Parsing.py]: I like sentences to end with a dot. :)

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -3207,9 +3146,6 @@ def declare_python_arg(self, env, arg):
     def analyse_expressions(self, env):
         self.local_scope.directives = env.directives
         self.analyse_default_values(env)
-        self.analyse_annotations(env)
-        if self.return_type_annotation:
-            self.return_type_annotation = self.analyse_annotation(env, self.return_type_annotation)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: Why is this no longer needed? Shouldn't the nodes still be analysed? It's up to the annotation node itself to decide to do nothing, not the node that contains it.

File: Cython/Compiler/ExprNodes.py [hunk 3]
```diff
@@ -1930,15 +1930,15 @@ def declare_from_annotation(self, env, as_target=False):
             return
 
         annotation = self.annotation
-        if annotation.is_string_literal:
+        if annotation.expr.is_string_literal:
             # name: "description" => not a type, but still a declared variable or attribute
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: I wonder if this is really still true. Shouldn't string literals behave just like any other type declaration now?

File: Cython/Compiler/ExprNodes.py [hunk 4]
```diff
@@ -13563,7 +13555,83 @@ def generate_result_code(self, code):
...
+            from .AutoDocTransforms import AnnotationWriter
+            string = StringEncoding.EncodedString(AnnotationWriter().write(expr))
+            string = StringNode(pos, unicode_value=string, value=string.as_utf8_string())
+        self.string = string
+        self.expr = expr
+
+    # def analyse_types doesn't exist because nothing needs doing
+
+    def coerce_to_pyobject(self, env):
+        return self.string
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: This is dangerous. Calling `coerce_to_pyobject()` is intended to be semantically like `coerce_to(py_object_type)`. Here, both would do entirely different things.

File: Cython/Compiler/Parsing.py (hunk 1 above)
Comment (McSinyx) [reply Cython/Compiler/Parsing.py]: I'm just wondering around, but does Cython follows PEP 257?

> Unless the entire docstring fits on a line, place the closing quotes on a line by themselves.

Comment (scoder) [reply Cython/Compiler/Parsing.py]: Cython's code base (more exactly, Pyrex's code base) is actually older than PEP-8 and PEP-257. :)
But yes, the triple quotes should be on their own next line, too.

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py]: I've added these back in (and in ExprNodes.py too) in a slightly simplified form.

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py]: Not sure... string literals are distinguishable from other types. E.g.

```def f(a: "something"):
    pass```

gets translated to an annotation of `'"something"'` (i.e. it's quoted inside the string literal).

So I don't automatically think it makes sense to change how Cython infers the types from the annotations - it has exactly the same information as before and all that's really changed is the representation in the `__annotations__` dictionary.

With that said, this looks like a silly and inconsistent special case, so maybe it should go. I haven't done it this commit though

```def f(a: float, b: 'float'):
    c: float = a
    d: 'float' = a
    e: float
    f: 'float'```

In the above `a`, `c`, `d`, `e` end up typed as a C double, `b` ends up as a C float, and `f` ends up as a PyObject so it's all a bit odd

**Comment by da-woods** [comment]
I've address the rest of the comments, apart from the string issue which is a bit odd. I could see it might be a good time to change the annotation->type translation, but it isn't explicitly linked with the PEP-563 changes

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py]: Don't know if it adds anything, but it's tested in https://github.com/cython/cython/blob/20787f28fae6c0c828940a25e580b56a1f594efd/tests/run/pep526_variable_annotations.py#L40

**Comment by da-woods** [comment]
I think the CI failure is just related to the stuff that's being fixed in https://github.com/cython/cython/pull/3291 and nothing to do with any changes. I'll close/re-open once that's fixed to rerun the tests.

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (scoder) [reply Cython/Compiler/ExprNodes.py]: Your example shows that it's definitely inconsistent currently. Given PEP-563, I think strings can be ignored now, because forward declarations are no longer a problem. However, they weren't (always) ignored before, so maybe we should show a warning for them if we detect a type in a string in 3.0, and really ignore them in 3.1?

File: Cython/Compiler/ExprNodes.py [hunk 5]
```diff
@@ -13563,7 +13558,81 @@ def generate_result_code(self, code):
...
+        return self.analyse_type_annotation(env)[1]
+
+    def analyse_type_annotation(self, env, assigned_value=None):
+        annotation = self.expr
+        base_type = None
+        is_ambiguous = False
+        explicit_pytype = explicit_ctype = False
+        if annotation.is_dict_literal:
+            warning(annotation.pos,
+                    "Dicts should no longer be used as type annotations. Use 'cython.int' etc. directly.")
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: I think Cython 3.0 is the right time to remove this. But that's not part of this PR.

File: Cython/Compiler/ExprNodes.py (hunk 3 above)
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py]: There was actually already a warning for annotations as strings (I'm struggling to work out how to link to it here, but in this commit it's in ExprNodes.py line 13625) so I've just increased the level so it shows by default.

File: Cython/Compiler/ExprNodes.py [hunk 6]
```diff
@@ -13563,7 +13558,81 @@ def generate_result_code(self, code):
...
+    # annotation is evaluated into a Python Object
+
+    subexprs = []
+    def __init__(self, pos, expr, string=None):
+        """string is expected to already be a StringNode or None"""
+        ExprNode.__init__(self, pos)
+        if string is None:
+            # import doesn't work at top of file?
+            from .AutoDocTransforms import AnnotationWriter
+            string = StringEncoding.EncodedString(AnnotationWriter().write(expr))
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py]: I've added a warning for these. I originally tried looking for ways to just copy the original code rather than regenerating it, but I couldn't work out a good way to do it. It's the sort of change that would be easy to make in the future though.

File: tests/run/cython3.pyx [hunk 7]
```diff
@@ -578,15 +578,15 @@ def annotation_syntax(a: "test new test", b : "other" = 2, *args: "ARGS", **kwar
     >>> len(annotation_syntax.__annotations__)
     5
     >>> print(annotation_syntax.__annotations__['a'])
-    test new test
+    u'test new test'
```
[the remark this answers is not in the document]
Comment (da-woods) [reply tests/run/cython3.pyx]: I don't think it does any harm. It's added by the code generation in `AnnotationWriter` (or it's base classes). The only real reason to change it would be that it generates a slightly different representation to Python, but since equivalent representations are OK in the PEP I don't think it's a big problem.

[the remark this answers is not in the document]
Comment (scoder) [reply tests/run/cython3.pyx]: It's not what users of Python 3.x expect, though. Cython will have to adapt to that at some point, and 3.0/1 seems a good time to do it, better now than in 3.2.
